### PR TITLE
fix(useDebouncedValue): types of useDebouncedValue will only add null to the union if initializeWithNull was passed

### LINKED ISSUE #1573: types for useDebouncedValue
I've spotted that the types for `useDebouncedValue` are not ideal, as they alway return a union with `null`, even if we don't add the `initializeWithNull` option. Consider:

```
const [debouncedValue] = useDebouncedValue('string', 200)
```

`debouncedValue` will be of type `string | null`, even though it should be `string`

[TypeScript playground reproduction](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwK4GcCmARNAjCSB2AxmgCYBqAhgDZJoC+cAZlBCHAOQsQDWK7AsACghhCPhTwA2iRx4ipSjTQBdOAF44qTLILFy1WgAp2EqMHwBzdgBo4AJgAMDgJRCA9G7hevAPQD8QA)

the `| null` should only be added if we explicitly provide option `initializeWithNull: true`:

```
const [debouncedValue] = useDebouncedValue('string', 200, { initializeWithNull: true })
```

I think this can be solved with conditional return types and another generic. Happy to contribute this :)

### PR DESCRIPTION
**Pull request by TkDodo**
fixes #1573

### CODE REVIEW

**Comment by TkDodo** [comment]
I haven't found any other type tests so I just added the helpers I usually work with

File: packages/rooks/src/__tests__/useDebouncedValue.spec.ts [hunk 1]
```diff
@@ -100,4 +110,42 @@ describe("useDebouncedValue", () => {
     });
     expect(result.current[0]).toBe(mockValue);
   });
+
+  describe("types", () => {
+    it("should return a union with null when initializeWithNull is true", () => {
+      doNotExecute(() => {
```
Comment (imbhargav5) [inline packages/rooks/src/__tests__/useDebouncedValue.spec.ts:116]: Holy! I love this pattern of just testing types without running code. Very nice!

File: packages/rooks/src/hooks/useDebouncedValue.ts [hunk 2]
```diff
@@ -4,17 +4,23 @@ import { useDebounce } from "./useDebounce";
...
 };
 
-type UseDebouncedValueReturnType<T> = [
-  debouncedValue: T | null,
-  immediatelyUpdateDebouncedValue: Dispatch<SetStateAction<T | null>>
+type UseDebouncedValueReturnType<
+  TValue = unknown,
+  TInitializeWithNull extends boolean = false
+> = [
+  debouncedValue: TInitializeWithNull extends true ? TValue | null : TValue,
```
Comment (imbhargav5) [inline packages/rooks/src/hooks/useDebouncedValue.ts:20]: Very nice!

**Comment by imbhargav5** [comment]
Love this @TkDodo ! Thank you very much!

**Pull request merged**